File: lc_py/lcPy/stack/nextGreaterElement.py
from typing import List


class Solution:
    def nextGreaterElement(self, nums1, nums2: List[int]) -> List[int]:
        idx = {x: i for i, x in enumerate(nums1)}

        st = []

        ans = [-1] * len(nums1)

        for x in nums2:
            while st and x > st[-1]:
                ans[idx[st.pop()]] = x
            if x in idx:
                st.append(x)
        return ans


if __name__ == '__main__':
    s = Solution()
    s.nextGreaterElement([4, 1, 2], [1, 3, 4, 2])

File: lc_go/lcGo/stack/nextGreaterElement.go
package main

func nextGreaterElement(nums1, nums2 []int) []int {
	ans := make([]int, len(nums1))
	idx := make(map[int]int, len(nums1))
	for i, x := range nums1 {
		idx[x] = i
	}
	for i, _ := range ans {
		ans[i] = -1
	}

	st := make([]int, len(nums2))

	for _, x := range nums2 {
		for len(st) > 0 && x > st[len(st)-1] {
			ans[idx[st[len(st)-1]]] = x
			st = st[:len(st)-1]
		}
		if _, ok := idx[x]; ok {
			st = append(st, x)
		}
	}
	return ans
}

func main() {

}
